source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.cc

```diff
@@ -0,0 +1,70 @@
+#include "source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.h"
+
+#include <dirent.h>
+#include <unistd.h>
+
+#include <array>
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <thread>
+
+#include "absl/strings/str_split.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ResourceMonitors {
+namespace CpuUtilizationMonitor {
+
+CpuUtilizationMonitor::CpuUtilizationMonitor(
+    const envoy::extensions::resource_monitors::cpu_utilization::v3::
+        CpuUtilizationConfig& /*config*/,
+    std::unique_ptr<CpuStatsReader> cpu_stats_reader)
+    : utilization_(0.0), cpu_stats_reader_(std::move(cpu_stats_reader)) {
+  previous_cpu_times_ = cpu_stats_reader_->getCpuTimes();
+}
+
+void CpuUtilizationMonitor::updateResourceUsage(Server::ResourceUpdateCallbacks& callbacks) {
+  CpuTimes cpu_times = cpu_stats_reader_->getCpuTimes();
+  if (!cpu_times.is_valid) {
+    ENVOY_LOG_MISC(error, "Can't open file to read CPU utilization");
+    const auto& error = EnvoyException("Can't open file to read CPU utilization");
+    callbacks.onFailure(error);
+    return;
+  }
+
+  const int64_t work_over_period = cpu_times.work_time - previous_cpu_times_.work_time;
+  const int64_t total_over_period = cpu_times.total_time - previous_cpu_times_.total_time;
+  if (work_over_period < 0 || total_over_period <= 0) {
+    ENVOY_LOG_MISC(error,
```

Comment:
likewise as above -- there's a failed update stat that gets incremented that can be used for monitoring the failures

Reply:
Removed.